risc0/zkvm/src/guest/env.rs

```diff
@@ -96,16 +97,41 @@ pub(crate) fn finalize() {
     }
 }
 
-/// Exchange data with the host, returning the data from the host
-/// as a slice of bytes.
-pub fn send_recv(channel: u32, buf: &[u8]) -> &'static [u8] {
-    unsafe { sys_io(channel, buf.as_ptr(), buf.len()) }
+/// Send data to the host.  The host may return two words back.
+pub fn send_host(channel: u32, buf: &[u8]) -> (u32, u32) {
+    unsafe { sys_io(&mut [] as _, 0, buf.as_ptr(), buf.len(), channel) }
 }
 
-/// Exchange data with the host, returning the data from the host as
-/// a slice of words and the length in bytes.
-pub fn send_recv_as_u32(channel: u32, buf: &[u8]) -> &'static [u32] {
-    bytemuck::cast_slice(send_recv(channel, buf))
+/// Exchange data with the host.
+pub fn send_recv_sized(
+    channel: u32,
+    to_host_buf: &[u8],
+    from_host_bytes: usize,
+) -> (u32, u32, &'static [u8]) {
+    unsafe {
+        let nwords = (from_host_bytes + WORD_SIZE - 1) / WORD_SIZE;
+        let from_host_buf = zkvm_abi_alloc_words(nwords);
```

Comment:
I wonder if we should let the user decide where the buffer to be filled comes from? Maybe they want to use their own custom allocator, or for it live in a local/stack variable?

Reply:
Yeah, pondering this for a bit I think this is kind of confusing and limiting also.

Anyways, I redid the public facing API parts of this PR and separated out "send_recv_raw" from "send_recv_slice".  I think this new way is cleaner and easier to understand and use; can you please take another look?